src/Util/JSON.php

```diff
@@ -50,4 +66,78 @@ public static function decode(string $data)
 
         return $decodedData;
     }
+
+    /**
+     * Performs sanity checks on data that shall be encoded to JSON.
+     *
+     * @param mixed $data     The data to sanitize
+     * @param int   $maxDepth The maximum depth to walk through `$data`
+     *
+     * @return mixed
+     *
+     * @throws JsonException
+     */
+    private static function sanitizeData($data, int $maxDepth)
+    {
+        if ($maxDepth < 0) {
+            throw new JsonException('Max depth limit of data to be sanitized reached.');
+        }
+
+        if (\is_string($data)) {
+            return self::convertStringToUtf8($data);
+        } elseif (\is_array($data) || \is_object($data)) {
+            $output = [];
+
+            foreach ($data as $key => $value) {
+                if (\is_string($key)) {
+                    $key = self::convertStringToUtf8($key);
+                }
+
+                if (\is_string($value)) {
+                    $value = self::convertStringToUtf8($value);
+                } elseif (\is_array($value) || \is_object($value)) {
+                    // This check is here because the `Event::toArray()` method
+                    // is broken and doesn't return all child items as scalars
+                    // or objects/arrays, so the sanitification would fail (e.g.
+                    // on breadcrumb objects which do not expose public properties
+                    // to iterate on)
+                    if (\is_object($value) && method_exists($value, 'toArray')) {
+                        $value = $value->toArray();
+                    }
+
+                    $value = self::sanitizeData($value, $maxDepth - 1);
+                }
+
+                $output[$key] = $value;
+            }
+
+            return \is_array($data) ? $output : (object) $output;
+        } else {
+            return $data;
+        }
+    }
+
+    /**
+     * Converts a string to UTF-8 to avoid errors during its encoding to
+     * the JSON format.
+     *
+     * @param string $value The text to convert to UTF-8
+     */
+    private static function convertStringToUtf8(string $value): string
+    {
+        $previousSubstituteCharacter = mb_substitute_character();
+        $encoding = mb_detect_encoding($value, mb_detect_order(), true);
+
+        mb_substitute_character(0xfffd);
+
+        if (false === $encoding) {
+            $value = mb_convert_encoding($value, 'UTF-8', 'UTF-8');
+        } else {
+            $value = mb_convert_encoding($value, 'UTF-8', $encoding);
```

Comment:
You could simplify this whole if with 
```suggestion
            $value = mb_convert_encoding($value, 'UTF-8', $encoding ?: 'UTF-8');
```